File: src/main/python/AoC2024_20.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 20
#

import multiprocessing
import os
import sys
from dataclasses import dataclass
from typing import Self

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.geometry import Direction
from aoc.geometry import Turn
from aoc.grid import CharGrid

TEST = """\
###############
#...#...#.....#
#.#.#.#.#.###.#
#S#...#.#.#...#
#######.#.#.###
#######.#.#...#
#######.#.###.#
###..E#...#...#
###.#######.###
#...###...#...#
#.#####.#.###.#
#.#...#.#.#...#
#.#.#.#.#.#.###
#...#...#...###
###############
"""

Cell = tuple[int, int]


@dataclass(frozen=True)
class RaceTrack:
    track: list[Cell]
    distances: list[int]
    h: int
    w: int

    @classmethod
    def from_input(cls, input_data: InputData) -> Self:
        grid = CharGrid.from_strings(list(input_data))
        h, w = grid.get_height(), grid.get_width()
        start = next(
            cell for cell in grid.get_cells() if grid.get_value(cell) == "S"
        )
        end = next(
            cell for cell in grid.get_cells() if grid.get_value(cell) == "E"
        )
        direction = next(
            d
            for d in Direction.capitals()
            if grid.get_value(start.at(d)) != "#"
        )
        pos = start
        dist = 0
        track = list[Cell]()
        distances = [sys.maxsize for r in range(h) for c in range(w)]
        while True:
            distances[pos.row * h + pos.col] = dist
            track.append((pos.row, pos.col))
            if pos == end:
                break
            direction = next(
                d
                for d in (
                    direction,
                    direction.turn(Turn.RIGHT),
                    direction.turn(Turn.LEFT),
                )
                if grid.get_value(pos.at(d)) != "#"
            )
            pos = pos.at(direction)
            dist += 1
        return cls(track, distances, h, w)

    def count_shortcuts(self, cheat_len: int, target: int) -> int:
        count = 0
        for r, c in self.track:
            d_cell = self.distances[r * self.h + c]
            for md in range(2, cheat_len + 1):
                min_req = target + md
                for dr in range(md + 1):
                    dc = md - dr
                    for rr, cc in {
                        (r + dr, c + dc),
                        (r + dr, c - dc),
                        (r - dr, c + dc),
                        (r - dr, c - dc),
                    }:
                        if not (0 <= rr < self.h and 0 <= cc < self.w):
                            continue
                        d_n = self.distances[rr * self.h + cc]
                        if d_n == sys.maxsize:
                            continue
                        if d_n - d_cell >= min_req:
                            count += 1
        return count


Input = RaceTrack
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return RaceTrack.from_input(input_data)

    def solve(self, racetrack: RaceTrack, cheat_len: int, target: int) -> int:
        ans = multiprocessing.Manager().dict()

        def count_shortcuts(pid: str, racetrack_i: RaceTrack) -> None:
            ans[pid] = racetrack_i.count_shortcuts(cheat_len, target)

        if sys.platform.startswith("win"):
            count_shortcuts("main", racetrack)
        else:
            n = os.process_cpu_count()
            cells: list[list[Cell]] = [[] for _ in range(n)]
            for cnt, cell in enumerate(racetrack.track):
                cells[cnt % n].append(cell)
            jobs = []
            for i in range(n):
                racetrack_i = RaceTrack(
                    cells[i],
                    racetrack.distances,
                    racetrack.h,
                    racetrack.w,
                )
                p = multiprocessing.Process(
                    target=count_shortcuts,
                    args=(str(i), racetrack_i),
                )
                jobs.append(p)
                p.start()
            for p in jobs:
                p.join()
        return sum(ans.values())

    def part_1(self, racetrack: Input) -> Output1:
        return self.solve(racetrack, 2, 100)

    def part_2(self, racetrack: Input) -> Output2:
        return self.solve(racetrack, 20, 100)

    def samples(self) -> None:
        inputs = self.parse_input(TEST.splitlines())
        assert self.solve(inputs, 2, 2) == 44
        assert self.solve(inputs, 20, 50) == 285


solution = Solution(2024, 20)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_20.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.geometry.Direction;
import com.github.pareronia.aoc.geometry.Turn;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2024_20 extends SolutionBase<CharGrid, Integer, Integer> {

    private AoC2024_20(final boolean debug) {
        super(debug);
    }

    public static AoC2024_20 create() {
        return new AoC2024_20(false);
    }

    public static AoC2024_20 createDebug() {
        return new AoC2024_20(true);
    }

    @Override
    protected CharGrid parseInput(final List<String> inputs) {
        return CharGrid.from(inputs);
    }

    private Integer solve(final CharGrid grid, final int cheatLength, final int target) {
        final Cell start = grid.getAllEqualTo('S').findFirst().orElseThrow();
        final Cell end = grid.getAllEqualTo('E').findFirst().orElseThrow();
        Direction dir =
                Direction.CAPITAL.stream()
                        .filter(d -> grid.getValue(start.at(d)) != '#')
                        .findFirst()
                        .orElseThrow();
        Cell pos = start;
        int dist = 0;
        final List<Cell> track = new ArrayList<>();
        final int h = grid.getHeight();
        final int[] d = new int[grid.getWidth() * h];
        Arrays.fill(d, Integer.MAX_VALUE);
        while (true) {
            d[pos.getRow() * h + pos.getCol()] = dist;
            track.add(pos);
            if (pos.equals(end)) {
                break;
            }
            final Cell p = pos;
            dir =
                    List.of(dir, dir.turn(Turn.RIGHT), dir.turn(Turn.LEFT)).stream()
                            .filter(dd -> grid.getValue(p.at(dd)) != '#')
                            .findFirst()
                            .orElseThrow();
            pos = pos.at(dir);
            dist += 1;
        }
        int ans = 0;
        for (final Cell cell : track) {
            final int dCell = d[cell.getRow() * h + cell.getCol()];
            for (int md = 2; md <= cheatLength; md++) {
                final int minReq = target + md;
                ans +=
                        cell.allAtManhattanDistance(md)
                                .filter(n -> grid.isInBounds(n))
                                .filter(
                                        n -> {
                                            final int dN = d[n.getRow() * h + n.getCol()];
                                            return dN != Integer.MAX_VALUE && dN - dCell >= minReq;
                                        })
                                .count();
            }
        }
        return ans;
    }

    @Override
    public Integer solvePart1(final CharGrid grid) {
        return solve(grid, 2, 100);
    }

    @Override
    public Integer solvePart2(final CharGrid grid) {
        return solve(grid, 20, 100);
    }

    @Override
    public void samples() {
        final AoC2024_20 test = AoC2024_20.createDebug();
        final CharGrid input = test.parseInput(StringOps.splitLines(TEST));
        assert test.solve(input, 2, 2) == 44;
        assert test.solve(input, 20, 50) == 285;
    }

    public static void main(final String[] args) throws Exception {
        AoC2024_20.create().run();
    }

    private static final String TEST =
            """
            ###############
            #...#...#.....#
            #.#.#.#.#.###.#
            #S#...#.#.#...#
            #######.#.#.###
            #######.#.#...#
            #######.#.###.#
            ###..E#...#...#
            ###.#######.###
            #...###...#...#
            #.#####.#.###.#
            #.#...#.#.#...#
            #.#.#.#.#.#.###
            #...#...#...###
            ###############
            """;
}
